File: account-service/src/main/java/com/wks/servicemarketplace/accountservice/config/GraphQLFactory.java
package com.wks.servicemarketplace.accountservice.config;

import com.wks.servicemarketplace.accountservice.adapters.graphql.AddressDataFetcher;
import com.wks.servicemarketplace.accountservice.adapters.graphql.CreateAddressDataFetcher;
import com.wks.servicemarketplace.accountservice.adapters.graphql.CreateCustomerDataFetcher;
import graphql.GraphQL;
import graphql.execution.instrumentation.ChainedInstrumentation;
import graphql.schema.GraphQLSchema;
import graphql.schema.idl.RuntimeWiring;
import io.gqljf.federation.FederatedSchemaBuilder;
import io.gqljf.federation.tracing.FederatedTracingInstrumentation;
import org.glassfish.hk2.api.Factory;

import javax.inject.Inject;
import java.io.InputStream;
import java.util.Collections;

public class GraphQLFactory implements Factory<GraphQL> {

    private GraphQL graphQL;

    @Inject
    public GraphQLFactory(CreateCustomerDataFetcher createCustomerDataFetcher,
                          CreateAddressDataFetcher createAddressDataFetcher,
                          AddressDataFetcher addressDataFetcher) {
        final InputStream schemaInputStream = getClass().getClassLoader().getResourceAsStream("schema.graphqls");

        final GraphQLSchema transformedGraphQLSchema = new FederatedSchemaBuilder()
                .schemaInputStream(schemaInputStream)
                .runtimeWiring(createRuntimeWiring(createCustomerDataFetcher, createAddressDataFetcher, addressDataFetcher))
                .excludeSubscriptionsFromApolloSdl(true)
                .build();
        this.graphQL = GraphQL.newGraphQL(transformedGraphQLSchema)
                .instrumentation(new ChainedInstrumentation(Collections.singletonList(new FederatedTracingInstrumentation())))
                .build();
    }

    private RuntimeWiring createRuntimeWiring(CreateCustomerDataFetcher createCustomerDataFetcher,
                                              CreateAddressDataFetcher createAddressDataFetcher, AddressDataFetcher addressDataFetcher) {
        return RuntimeWiring.newRuntimeWiring()
                .type("Query", builder -> builder.dataFetcher("address", addressDataFetcher))
                .type("Mutation", builder ->
                        builder.dataFetcher("createCustomer", createCustomerDataFetcher)
                                .dataFetcher("createAddress", createAddressDataFetcher))
                .build();
    }

    @Override
    public GraphQL provide() {
        return graphQL;
    }

    @Override
    public void dispose(GraphQL instance) {
    }
}

File: auth-service/src/main/kotlin/com/wks/servicemarketplace/authservice/config/GraphQLFactory.kt
package com.wks.servicemarketplace.authservice.config

import com.wks.servicemarketplace.authservice.adapters.graphql.ApiTokenDataFetcher
import com.wks.servicemarketplace.authservice.adapters.graphql.GraphQLDataFetcherExceptionHandler
import com.wks.servicemarketplace.authservice.adapters.graphql.LoginDataFetcher
import com.wks.servicemarketplace.authservice.adapters.graphql.RegisterDataFetcher
import graphql.GraphQL
import graphql.execution.AsyncExecutionStrategy
import graphql.execution.instrumentation.ChainedInstrumentation
import graphql.schema.idl.RuntimeWiring
import io.gqljf.federation.FederatedSchemaBuilder
import io.gqljf.federation.tracing.FederatedTracingInstrumentation
import org.glassfish.hk2.api.Factory
import javax.inject.Inject

class GraphQLFactory @Inject constructor(private val loginDataFetcher: LoginDataFetcher,
                                         private val registerDataFetcher: RegisterDataFetcher,
                                         private val apiTokenFetcher: ApiTokenDataFetcher) : Factory<GraphQL> {

    private val graphQL: GraphQL

    init {
        val schemaInputStream = javaClass.classLoader.getResourceAsStream("schema.graphqls")

        val transformedGraphQLSchema = FederatedSchemaBuilder()
                .schemaInputStream(schemaInputStream)
                .runtimeWiring(createRuntimeWiring())
                .excludeSubscriptionsFromApolloSdl(true)
                .build()
        graphQL = GraphQL.newGraphQL(transformedGraphQLSchema)
                .queryExecutionStrategy(AsyncExecutionStrategy(GraphQLDataFetcherExceptionHandler()))
                .mutationExecutionStrategy(AsyncExecutionStrategy(GraphQLDataFetcherExceptionHandler()))
                .instrumentation(ChainedInstrumentation(listOf(FederatedTracingInstrumentation())))
                .build()
    }

    private fun createRuntimeWiring(): RuntimeWiring? {
        return RuntimeWiring.newRuntimeWiring()
                .type("Mutation") { builder ->
                    builder.dataFetcher("signIn", loginDataFetcher)
                            .dataFetcher("signUp", registerDataFetcher)
                            .dataFetcher("apiToken", apiTokenFetcher)
                }
                .build()
    }

    override fun provide() = graphQL

    override fun dispose(instance: GraphQL?) {}
}